src/server/session.ts

```diff
@@ -1264,57 +1275,123 @@ export class Session<TMessage = string> implements EventSender {
         this.sendDiagnosticsEvent(file, project, project.getLanguageService().getSuggestionDiagnostics(file), "suggestionDiag");
         tracing?.pop();
     }
 
-    private sendDiagnosticsEvent(file: NormalizedPath, project: Project, diagnostics: readonly Diagnostic[], kind: protocol.DiagnosticEventKind): void {
+    private regionSemanticCheck(file: NormalizedPath, project: Project, ranges: TextRange[]): boolean {
+        this.diagnosticsTime = this.hrtime();
+        tracing?.push(tracing.Phase.Session, "regionSemanticCheck", { file, configFilePath: (project as ConfiguredProject).canonicalConfigFilePath }); // undefined is fine if the cast fails
+        let diagnosticsResult;
+        if (!this.shouldDoRegionCheck(file) || !(diagnosticsResult = project.getLanguageService().getRegionSemanticDiagnostics(file, ranges))) {
+            tracing?.pop();
+            return false;
+        }
+        this.sendDiagnosticsEvent(file, project, diagnosticsResult.diagnostics, "regionSemanticDiag", diagnosticsResult.spans);
+        tracing?.pop();
+        return true;
+    }
+
+    // We should only do the region-based semantic check if we think it would be considerably faster than a whole-file semantic check
+    protected shouldDoRegionCheck(file: NormalizedPath): boolean {
+        const perf = this.semanticCheckPerformance.get(file);
+        return !!perf && perf > 1000;
+    }
+
+    private sendDiagnosticsEvent(file: NormalizedPath, project: Project, diagnostics: readonly Diagnostic[], kind: protocol.DiagnosticEventKind, spans?: TextSpan[]): void {
         try {
-            this.event<protocol.DiagnosticEventBody>({ file, diagnostics: diagnostics.map(diag => formatDiag(file, project, diag)) }, kind);
+            const scriptInfo = Debug.checkDefined(project.getScriptInfo(file));
+            const duration = hrTimeToMilliseconds(this.hrtime(this.diagnosticsTime));
+            if (kind === "semanticDiag") {
+                this.semanticCheckPerformance?.set(file, duration);
+            }
+
+            const body: protocol.DiagnosticEventBody = {
+                file,
+                diagnostics: diagnostics.map(diag => formatDiag(file, project, diag)),
+                spans: spans?.map(span => toProtocolTextSpan(span, scriptInfo)),
+            };
+            if (this.includeDiagnosticsDuration) {
+                body.duration = duration;
+            }
+            this.event<protocol.DiagnosticEventBody>(
+                body,
+                kind,
+            );
         }
         catch (err) {
             this.logError(err, kind);
         }
     }
 
     /** It is the caller's responsibility to verify that `!this.suppressDiagnosticEvents`. */
-    private updateErrorCheck(next: NextStep, checkList: readonly string[] | readonly PendingErrorCheck[], ms: number, requireOpen = true) {
+    private updateErrorCheck(

```

Comment:
Please update updateErrorCheck to be more readable as:
```suggestion
private updateErrorCheck(next: NextStep, checkList: readonly string[] | readonly PendingErrorCheck[], ms: number, requireOpen = true) {
        Debug.assert(!this.suppressDiagnosticEvents); // Caller's responsibility
        const seq = this.changeSeq;
        const followMs = Math.min(ms, 200);
        let index = 0;
        const getFullFileIndex = () => index - checkList.length;
        let filesFullCheck: PendingErrorCheck[] | undefined;
        const goNext = () => {
            index++;
            if (checkList.length > index || (filesFullCheck?.length ?? 0) > getFullFileIndex()) {
                next.delay("checkOne", followMs, checkOne);
            }
        };
        const checkOne = () => {
            if (this.changeSeq !== seq) {
                return;
            }
            let item: string | PendingErrorCheck | undefined = checkList[index] ?? filesFullCheck![getFullFileIndex()];
            if (isString(item)) {
                // Find out project for the file name
                item = this.toPendingErrorCheck(item);
                if (!item) {
                    // Ignore file if there is no project for the file
                    goNext();
                    return;
                }
            }
            const { fileName, project, ranges, skipSyntax } = item;
            // Ensure the project is up to date before checking if this file is present in the project.
            updateProjectIfDirty(project);
            if (!project.containsFile(fileName, requireOpen)) {
                return;
            }
            if (!skipSyntax) {
                this.syntacticCheck(fileName, project);
                if (this.changeSeq !== seq) {
                    return;
                }
            }
            // Don't provide semantic diagnostics unless we're in full semantic mode.
            if (project.projectService.serverMode !== LanguageServiceMode.Semantic) {
                goNext();
                return;
            }
            if (ranges) {
                return next.immediate("regionSemanticCheck", () => {
                    const didRegionCheck = this.regionSemanticCheck(fileName, project, ranges);
                    if (didRegionCheck) {
                        (filesFullCheck ??= []).push({ fileName, project, skipSyntax: true });
                        goNext();
                        return;
                    }
                    // If we opted out of region check, do semantic check as normal in next step
                });
            }
            next.immediate("semanticCheck", () => {
                this.semanticCheck(fileName, project);
                if (this.changeSeq !== seq) {
                    return;
                }
                if (this.getPreferences(fileName).disableSuggestions) {
                    goNext();
                    return;
                }
                next.immediate("suggestionCheck", () => {
                    this.suggestionCheck(fileName, project);
                    goNext();
                });
            });
        };
        if ((checkList.length + (filesFullCheck?.length ?? 0)) > index && this.changeSeq === seq) {
            next.delay("checkOne", ms, checkOne);
        }
    }
```

Reply:
I'm a bit confused, what specifically about `updateErrorCheck` do you think could be more readable?

Reply:
Update: I changed the behavior of `updateErrorCheck` and as a consequence the code is simpler now, let me know if there are still any parts that are confusing or could be improved.